### PR TITLE
Add git revision option

### PR DESCRIPTION
**Pull request by shay-stratoscale**
@adir-stratoscale @eran-stratoscale @ravid-stratoscale

### CODE REVIEW

File: tests/test_cli.py [hunk 1]
```diff
@@ -86,6 +86,16 @@
     'workdir': 'test-workdir'
 }
 
+SKIPPER_CONF_WITH_GIT_REV = {
+    'registry': REGISTRY,
+    'build-container-image': SKIPPER_CONF_BUILD_CONTAINER_IMAGE,
+    'build-container-tag': SKIPPER_CONF_BUILD_CONTAINER_TAG,
```
Comment (eran-stratoscale) [inline tests/test_cli.py]: No need for it

File: tests/test_cli.py [hunk 2]
```diff
@@ -86,6 +86,16 @@
...
+    'make': {
+        'makefile': SKIPPER_CONF_MAKEFILE,
+    },
+    'use_git_revision': 'true'
```
Comment (eran-stratoscale) [inline tests/test_cli.py]: Perhaps use a name that better represent that parameter is relevant for the build container tag.

Comment (shay-stratoscale) [inline tests/test_cli.py]: Any suggestions?

**Comment by shay-stratoscale** [comment]
Review

File: skipper/cli.py [hunk 3]
```diff
@@ -25,8 +25,9 @@ def cli(ctx, registry, build_container_image, build_container_tag, build_contain
 
     ctx.obj['registry'] = registry
     ctx.obj['build_container_image'] = build_container_image
-    ctx.obj['build_container_tag'] = build_container_tag
     ctx.obj['build_container_net'] = build_container_net
+    ctx.obj['git_revision'] = build_container_tag == 'git:revision'
```
Comment (adir-stratoscale) [inline skipper/cli.py]: why do you need to propagate this value to commands functions? 
you can simply set `build_container_tag` with the current git revision, if `build_container_tag == git:revision`:
```python 
ctx.obj['build_container_tag'] = git.get_hash() if build_container_tag == 'git:revision' else build_container_tag
```

File: skipper/cli.py [hunk 4]
```diff
@@ -241,11 +248,17 @@ def _prepare_build_container(registry, image, tag):
...
+
+    else:
+        image_name = image
+        utils.logger.info("No build container tag was provided")
 
-    utils.logger.info("No build container tag was provided. Building from scratch...")
     dockerfile = utils.image_to_dockerfile(image)
-    command = ['docker', 'build', '-t', image, '-f', dockerfile, '.']
+    utils.logger.info("Building using docker file: %(dockerfile)s", dict(dockerfile=dockerfile))
+    command = ['docker', 'build', '-t', image_name, '-f', dockerfile, '.']
```
Comment (adir-stratoscale) [inline skipper/cli.py]: why? if we build the build container, lets tag it with latest.

Comment (adir-stratoscale) [inline skipper/cli.py]: @eran-stratoscale what do you think?

File: skipper/utils.py [hunk 5]
```diff
@@ -47,6 +48,10 @@ def remote_image_exist(registry, image, tag):
     requests.packages.urllib3.disable_warnings()
     url = IMAGE_TAGS_URL % dict(registry=registry, image=image)
     response = requests.get(url=url, verify=False)
+
+    if response.status_code != httplib.OK:
+        return False
+
```
Comment (adir-stratoscale) [inline skipper/utils.py]: :+1:

File: skipper/cli.py [hunk 6]
```diff
@@ -229,24 +233,36 @@ def version():
...
+        tagged_image_name = image
+        utils.logger.info("No build container tag was provided")
+
+    docker_file = utils.image_to_dockerfile(image)
+    utils.logger.info("Building image using docker file: %(docker_file)s", dict(docker_file=docker_file))
+    runner.run(['docker', 'build', '-t', image, '-f', docker_file, '.'])
+
+    if git_revision and not git.uncommitted_changes():
+        utils.logger.info("Tagging image with git revision: %(tag)s", dict(tag=tag))
+        runner.run(['docker', 'tag', image + ':latest', tagged_image_name])
```
Comment (adir-stratoscale) [inline skipper/cli.py]: not critical, but you could drop the ":latest"

File: tests/test_cli.py [hunk 7]
```diff
@@ -980,6 +991,23 @@ def test_run_with_defaults_from_config_file_including_workdir(self, skipper_runn
                                                         interactive=False, net='host', volumes=None,
                                                         workdir='test-workdir')
 
+    @mock.patch('__builtin__.open', create=True)
+    @mock.patch('os.path.exists', autospec=True, return_value=True)
+    @mock.patch('yaml.load', autospec=True, return_value=SKIPPER_CONF_WITH_GIT_REV)
+    @mock.patch('subprocess.check_output', autospec=True, return_value='1234567\n')
```
Comment (adir-stratoscale) [inline tests/test_cli.py]: i think you don't check all cases. for exampe, `git_version == True and uncommitted_changes == True`.

File: skipper/git.py [hunk 8]
```diff
@@ -6,4 +7,13 @@ def get_hash(short=False):
...
+
+    if uncommitted_changes():
+        logging.warning("*** Uncommitted changes present - Build container version might be outdated ***")
+
     return subprocess.check_output(git_command).strip()
+
+
+def uncommitted_changes():
+    """Return True is there are uncommitted changes."""
+    return subprocess.call(['git', 'diff-index', '--quiet', 'HEAD', '--']) != 0
```
Comment (shay-stratoscale) [inline skipper/git.py]: It will raise exception on a in case the return code is not zero

File: tests/test_cli.py (hunk 7 above)
Comment (shay-stratoscale) [inline tests/test_cli.py]: I'll add the test

**Comment by shay-stratoscale** [comment]
@ravid-stratoscale

**Pull request merged**